File: runners/csharp/Kikker.Billing.IntegrationTests/BillingTestOptions.cs
using System.Globalization;

namespace Kikker.Billing.IntegrationTests;

public sealed class BillingTestOptions
{
    public string RepositoryRoot { get; init; } = string.Empty;
    public string ComposeFile { get; init; } = string.Empty;
    public string EnvFile { get; init; } = string.Empty;
    public string AutoBillingBaseUrl { get; init; } = string.Empty;
    public string ContractNumber { get; init; } = string.Empty;
    public string FromDate { get; init; } = string.Empty;
    public string UntilDate { get; init; } = string.Empty;
    public string MySqlRootPassword { get; init; } = string.Empty;
    public TimeSpan WaitTimeout { get; init; }

    public Uri AutoBillingRunUri => new($"{AutoBillingBaseUrl.TrimEnd('/')}/auto-billing/run");

    public static BillingTestOptions Load()
    {
        var repositoryRoot = ResolveRepositoryRoot();
        var timeoutSeconds = ReadInt("KIKKER_ITEST_WAIT_TIMEOUT_SECONDS", 180);

        return new BillingTestOptions
        {
            RepositoryRoot = repositoryRoot,
            ComposeFile = Path.Combine(repositoryRoot, ReadString("KIKKER_ITEST_COMPOSE_FILE", "docker-compose.itest.yml")),
            EnvFile = Path.Combine(repositoryRoot, ReadString("KIKKER_ITEST_ENV_FILE", ".env.itest")),
            AutoBillingBaseUrl = ReadString("KIKKER_ITEST_AUTOBILLING_BASE_URL", "http://localhost:8085"),
            ContractNumber = ReadString("KIKKER_ITEST_CONTRACT_NUMBER", "C20241292"),
            FromDate = ReadString("KIKKER_ITEST_FROM", "2025-02-09"),
            UntilDate = ReadString("KIKKER_ITEST_UNTIL", "2026-02-09"),
            MySqlRootPassword = ReadString("KIKKER_ITEST_MYSQL_ROOT_PASSWORD", "kikker"),
            WaitTimeout = TimeSpan.FromSeconds(timeoutSeconds)
        };
    }

    private static string ResolveRepositoryRoot()
    {
        var configured = Environment.GetEnvironmentVariable("KIKKER_ITEST_ROOT");
        if (!string.IsNullOrWhiteSpace(configured))
        {
            return Path.GetFullPath(configured);
        }

        var current = new DirectoryInfo(AppContext.BaseDirectory);
        while (current is not null)
        {
            var composePath = Path.Combine(current.FullName, "docker-compose.itest.yml");
            if (File.Exists(composePath))
            {
                return current.FullName;
            }

            current = current.Parent;
        }

        throw new InvalidOperationException("Could not locate repository root from test runner output path.");
    }

    private static string ReadString(string key, string fallback)
    {
        var value = Environment.GetEnvironmentVariable(key);
        return string.IsNullOrWhiteSpace(value) ? fallback : value;
    }

    private static int ReadInt(string key, int fallback)
    {
        var value = Environment.GetEnvironmentVariable(key);
        return int.TryParse(value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var parsed)
            ? parsed
            : fallback;
    }
}

File: runners/java/src/test/java/com/kikker/billing/integrationtests/BillingTestOptions.java
package com.kikker.billing.integrationtests;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;

public final class BillingTestOptions {
    private final Path repositoryRoot;
    private final Path composeFile;
    private final Path envFile;
    private final String autoBillingBaseUrl;
    private final String contractNumber;
    private final String fromDate;
    private final String untilDate;
    private final String mysqlRootPassword;
    private final Duration waitTimeout;

    private BillingTestOptions(
        Path repositoryRoot,
        Path composeFile,
        Path envFile,
        String autoBillingBaseUrl,
        String contractNumber,
        String fromDate,
        String untilDate,
        String mysqlRootPassword,
        Duration waitTimeout
    ) {
        this.repositoryRoot = repositoryRoot;
        this.composeFile = composeFile;
        this.envFile = envFile;
        this.autoBillingBaseUrl = autoBillingBaseUrl;
        this.contractNumber = contractNumber;
        this.fromDate = fromDate;
        this.untilDate = untilDate;
        this.mysqlRootPassword = mysqlRootPassword;
        this.waitTimeout = waitTimeout;
    }

    public static BillingTestOptions load() {
        Path repositoryRoot = resolveRepositoryRoot();
        int timeoutSeconds = readInt("KIKKER_ITEST_WAIT_TIMEOUT_SECONDS", 180);

        return new BillingTestOptions(
            repositoryRoot,
            repositoryRoot.resolve(readString("KIKKER_ITEST_COMPOSE_FILE", "docker-compose.itest.yml")).normalize(),
            repositoryRoot.resolve(readString("KIKKER_ITEST_ENV_FILE", ".env.itest")).normalize(),
            readString("KIKKER_ITEST_AUTOBILLING_BASE_URL", "http://localhost:8085"),
            readString("KIKKER_ITEST_CONTRACT_NUMBER", "C20241292"),
            readString("KIKKER_ITEST_FROM", "2025-02-09"),
            readString("KIKKER_ITEST_UNTIL", "2026-02-09"),
            readString("KIKKER_ITEST_MYSQL_ROOT_PASSWORD", "kikker"),
            Duration.ofSeconds(timeoutSeconds)
        );
    }

    public Path repositoryRoot() {
        return repositoryRoot;
    }

    public Path composeFile() {
        return composeFile;
    }

    public Path envFile() {
        return envFile;
    }

    public String autoBillingBaseUrl() {
        return autoBillingBaseUrl;
    }

    public String contractNumber() {
        return contractNumber;
    }

    public String fromDate() {
        return fromDate;
    }

    public String untilDate() {
        return untilDate;
    }

    public String mysqlRootPassword() {
        return mysqlRootPassword;
    }

    public Duration waitTimeout() {
        return waitTimeout;
    }

    public String autoBillingRunUrl() {
        return autoBillingBaseUrl.replaceAll("/+$", "") + "/auto-billing/run";
    }

    private static Path resolveRepositoryRoot() {
        String configuredRoot = System.getenv("KIKKER_ITEST_ROOT");
        if (configuredRoot != null && !configuredRoot.isBlank()) {
            return Path.of(configuredRoot).toAbsolutePath().normalize();
        }

        Path current = Path.of("").toAbsolutePath().normalize();
        while (current != null) {
            if (Files.exists(current.resolve("docker-compose.itest.yml"))) {
                return current;
            }
            current = current.getParent();
        }

        throw new IllegalStateException("Could not locate repository root from the current working directory.");
    }

    private static String readString(String key, String fallback) {
        String value = System.getenv(key);
        return value == null || value.isBlank() ? fallback : value;
    }

    private static int readInt(String key, int fallback) {
        String value = System.getenv(key);
        if (value == null || value.isBlank()) {
            return fallback;
        }

        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException ignored) {
            return fallback;
        }
    }
}
